From 36073592a3ffe82bd20fba69cbd0625637aa6fe7 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 04 Jun 2019 19:56:46 +0300
Subject: [PATCH] [SW-4185] gaudi: pll: extend the gated clk switch

The hl_pll_hbw_gated_clk_sw() function is hard coded to work only on
the PLL divider output 0.
Extend it to work on all PLL dividers outputs.
This is achieved by generating the gate mask according to the PLL
configuration and switching clock output affected by the clock switch
glitch.

Change-Id: I8939521c69bc7dbd1aad84262d231661f21bbf65
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/asic.c b/gaudi/asic.c
index 8932694..e09a795 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -341,12 +341,34 @@
 		preboot_puts("Surprise! Woke up from WFE...\n");
 }
 
+static u32 hl_pll_build_gate_mask(struct hl_pll_cfg *cfg)
+{
+	int i, gmask = 0;
+
+	for (i = 0; i < HL_PLL_DIV_SIZE; i++)
+		if (cfg->div_wa[i] == HL_PLL_WA_INIT)
+			gmask |= 1 << i;
+
+	return ~gmask & 0xF;
+}
+
 static void hl_pll_hbw_gated_clk_sw(struct hl_pll_cfg *cfg)
 {
-	regmap_writel(0xE, cfg->base + HL_PLL_CLK_GATER);
+	int i;
+	u32 gmask = hl_pll_build_gate_mask(cfg);
+	u64 reg;
+
+	regmap_writel(gmask, cfg->base + HL_PLL_CLK_GATER);
 	hl_pll_udelay(1);
-	regmap_writel(HL_PLL_PLL_CLK, cfg->base + HL_PLL_DIV_SEL);
+
+	for (i = 0; i < HL_PLL_DIV_SIZE; i++) {
+		if (cfg->div_wa[i] == HL_PLL_WA_INIT) {
+			reg = cfg->base + HL_PLL_DIV_SEL + (i * sizeof(u32));
+			regmap_writel(cfg->div_sel[i], reg);
+		}
+	}
 	hl_pll_udelay(10);
+
 	regmap_writel(0xF, cfg->base + HL_PLL_CLK_GATER);
 	hl_pll_udelay(10);
 }
